### ISSUE TITLE
Porting t.w._newclient and t.w.t.test_newclient to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7485]
|<img alt="maximlacrima's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| maximlacrima reported|
|-|-|
|Trac ID|trac#7485|
|Type|enhancement|
|Created|2014-06-05 21:15:33Z|
|Branch|https://github.com/twisted/twisted/tree/newclient-py3-7485|

I have prepared a patch for `twisted.web._newclient` and `twisted.web.test.test_newclient`.

## Checklist from wiki/plan/python3

**1. Tests pass under 3.3.**

  Yes

**2. Ported modules were added to list of ported modules in twisted/python/dist3.py**

  Yes

**3. No warnings when running using -3 mode under Python 2.7.**

  There are two deprecation warnings for `python -3 bin/trial twisted.web.test.test_newclient`:

  ```
    /home/maxim/twistedtrunk/twisted/python/modules.py:291: DeprecationWarning: Overriding __eq__ blocks inheritance of __hash__ in 3.x
  class PythonModule(_ModuleIteratorHelper):
    /home/maxim/twistedtrunk/twisted/internet/address.py:169: DeprecationWarning: Overriding __eq__ blocks inheritance of __hash__ in 3.x
  ```

  But these warnings are not for `t.w._newclient`. `t.i.address` is marked as ported in `t.p.dist3`. And `t.p.modules` is not present in the dependency graph that I built for `t.w._newclient` and `t.w.t.test_newclient` (may it be used by trial?).  So I am not sure what to do about these warnings

**4. All non-docstring strings are clearly marked as one of the various string categories (bytes/unicode/identifier/version-dependent).**

  String literals that need to be `bytes` are prepended with `b`. Constants and strings that are passed to log messages and exceptions are of "text type" (str in Python 3). I didn't prepend "text type" with `u` (should I?).

**5. No use of 'str' except for things that really can differ across versions of Python.**

  `str` is not used.

'''6. from `__future__ import division, absolute_import` was added to the module.'''

  Division operator is not used in ported modules, so imported only `absolute_import`.

**7. Verify dictionary methods were ported correctly.**

  There are no method calls on dictionaries that would require porting.

**8. All calls to open() files should be in binary mode.**

  There are no calls to `open()`.

'''9. Unavoidable incompatibilities between Twisted APIs on Python 3 and the original behavior on Python 2 are documented in `docs/projects/core/howto/python3.rst`'''

  * `Failure.trap()`` is not used in `t.w._newclient`. Calls to ``Failure.trap()`` in `t.w.test.test_newclient` do not require code changes, because there are no checks for exception/failure types in case ``trap()`` raises an exception/failure.

  * Arguments, passed to log functions, are text strings.

'''10. Python 3 changes from Python 2 are addressed, including but not limited to:'''

  There are following changes changes throughout these two modules:

  1. `except ... as ...:`: line 398 in `_newclient`.

  2. ``nonlocal`` is a reserved word in Python 3. So change a variable name in lines 1692, 1694 in `test_newclient`.

  3. String formatting (`%`-operator) operations were wrapped into `compat.networkString` or replaced with `bytes.join` or concatanation (depending on what is more appropriate in a given case). See lines 638, 645, 681, 1216 in `_newclient` and line 244 in `test_client`.

  4. Use `@implementer` instead of `implements`.

  5. All other changes are explicitly marking some string literals as `bytes`.

'''11. twisted.python.compat compatibility functions are used, rather than reinventing the wheel.'''

  ``networkString`` is used in two places: lines 681 and 1216.

'''12. _PY3 checks are ideally only done at top of module, e.g. to define alternate helper modules.'''

  There are no ``_PY3`` checks.

== Notes ==

1. Docstrings were changed to reflect type changes, e.g. `"@type status: C{str}"` -> `"@type status: C{bytes}"`.

2. Fixed minor typos in code and docstrings.

3. Lines that were longer than 79 columns, have been adjusted to meet the standard.

Attachments:

* [_newclient.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b3d/b3dcdd454daca9b55accf9f64cc46832d41b7cef/27606a8cafb6ff3500a192f04bda6211c62c4224.patch) (79366 bytes) - added by maximlacrima on 2014-06-05 21:17:24Z - patch for t.w._newclient and t.w.t.test_newclient
* [_newclient.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b3d/b3dcdd454daca9b55accf9f64cc46832d41b7cef/e93d6cb86e6ba001c4f5f7341e80e220cd0bc930.patch) (79365 bytes) - added by maximlacrima on 2014-06-05 21:22:04Z - 
* [_newclient.3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b3d/b3dcdd454daca9b55accf9f64cc46832d41b7cef/bdc045dd167fd3545cd6065634167b6eb5053f69.patch) (107462 bytes) - added by maximlacrima on 2014-06-15 18:26:07Z - 
* [_newclient.4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b3d/b3dcdd454daca9b55accf9f64cc46832d41b7cef/442beea3cf34438485a36cfe98155bf1fe81a99d.patch) (98958 bytes) - added by maximlacrima on 2014-10-01 08:44:24Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7485 7485
type__enhancement enhancement
reporter__maximlacrima maximlacrima
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_newclient_py3_7485 branches/newclient-py3-7485
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1402002933000000 1402002933000000
changetime__1421485882311364 1421485882311364
version__None None
owner__glyph glyph
cc__jknight cc__maximlacrima
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="maximlacrima's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|maximlacrima commented|
|-|-|

Please delete the first attachment (I can't figure out how to delete attachments)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban set owner to maximlacrima|
|-|-|

Changes look good.

Many thanks for the great review request!

---------

Is ok to leave text variables without 'u' as long as the code works.

-------

```
- nonlocal = {'cancelled': False}
+ nonlocal_ = {'cancelled': False}
}}

Maybe you can find a better not to reduce the confusion with
reserved word.
I don't like the underline at the end.

------

You need to add a news file as described here
http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles

That file is used to generated to release report.

Thanks```{
- nonlocal = {'cancelled': False}
+ nonlocal_ = {'cancelled': False}
}}

Maybe you can find a better not to reduce the confusion with
reserved word.
I don't like the underline at the end.

------

You need to add a news file as described here
http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles

That file is used to generated to release report.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

> Is ok to leave text variables without 'u' as long as the code works.

I think it's better to add the `u`.  This indicates that someone has considered the type of the string and made a decision.  Relying on the native string type being correct gives no such indication.  It's also better if APIs only one of bytes or unicode, not both (or a different one depending on the Python version).  Some APIs have to be exceptions to this because of considerations beyond our control (for example, decisions made about how the stdlib changed in Python 3).  We should try to avoid introducing our own such difficulties when possible though.

**Comment by twisted-trac** [comment]
|<img alt="maximlacrima's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|maximlacrima commented|
|-|-|

As for your comments there are new changes in the patch:

* Added news file
* `nonlocal_` variable changed to `nonLocal`
* each string has been explicitly marked as a particular type, except:
    * docstrings
    * strings passed to `networkString` function, because it expects the argument in Python 3 to be of "text type" and in Python 2 -- bytes type.
    * strings passed as `msg` argument to `TestCase.assert*` methods.

**Comment by twisted-trac** [comment]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|wolfgang61 commented|
|-|-|

* shouldn't the news file be in twisted/web/topfiles?
* Why must HTTParser States  be unicode instead of bytes? Or would it be possible to just define STATUS, HEADER, BODY, DONE = range(4)? Python2 tests survive this. So either this is possible or more tests are needed. No need for unicode, I think.
* Please create separate patch files: At least one for whitespace changes only, one for fixes independent of PY3 (like test_newclient around line 610 or changed comments further down), and one with only real PY3 changes. This one should be as small as possible.
* I do not think getattr() should get the attribute name as unicode. Python2 will have to convert it to str and fail if it cannot do so. Try getattr(a,u'ü'). I understand that it is better to know that somebody made a conscious decision but this slows the execution.
* Same for DEFERRED_CLOSE and friends. No need to be unicode. Their only use is to be converted into method names, so Python2 will encode them to ascii.
* Please define @type for all _state attributes. @type C{str} if you define them as native strings (see DEFERRED_CLOSE). I think this @type should be enough to show there was a conscious decision.
* Values for producerState: This is just an enum, no need for unicode (see HTTParser State). Plus those values are defined in test/proto_helpers.py, and you should adhere to proto_helpers.py. Sure u'a' == 'a' returns True even for Python2 but again this slows execution.

**Comment by twisted-trac** [comment]
|<img alt="maximlacrima's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|maximlacrima commented|
|-|-|

Replying to [wolfgang61](#note_8):
> * shouldn't the news file be in twisted/web/topfiles?

I don't remember why exactly I put it in 'twisted/topfiles' instead of 'twisted/web/topfiles'. I think because I found other news files related to `twisted.web` in root topfiles directory. I will move it.

> * Why must HTTParser States  be unicode instead of bytes? Or would it be possible to just define STATUS, HEADER, BODY, DONE = range(4)? Python2 tests survive this. So either this is possible or more tests are needed. No need for unicode, I think.
> * I do not think getattr() should get the attribute name as unicode. Python2 will have to convert it to str and fail if it cannot do so. Try getattr(a,u'ü'). I understand that it is better to know that somebody made a conscious decision but this slows the execution.
> * Same for DEFERRED_CLOSE and friends. No need to be unicode. Their only use is to be converted into method names, so Python2 will encode them to ascii.
> * Please define @type for all _state attributes. @type C{str} if you define them as native strings (see DEFERRED_CLOSE). I think this @type should be enough to show there was a conscious decision.
> * Values for producerState: This is just an enum, no need for unicode (see HTTParser State). Plus those values are defined in test/proto_helpers.py, and you should adhere to proto_helpers.py. Sure u'a' == 'a' returns True even for Python2 but again this slows execution.

_newclient.2.patch​ is a patch that I think adheres to your points. That is, text/bytes literals are not explicitly prefixed with `b` or `u` as long as a particular type is not enforced by twisted API and the code continues to work with native strings in both Python 2 and 3. Tests were carefully changed to reflect this idea (e.g. headers, urls etc. must be bytes, but for many things native string type works fine), and in this way `t.w._newclient` had been ported to Python 3.

Then exarkun commented on the ticket, see comment:5, telling that relying on the native string type gives no indication that someone considered a particular type of the string, and also it is better when API supports exactly one type of the string in both versions of Python. I.e. if native string (text) is expected in Python 3, then unicode must be expected in Python 2. If bytes are expected in Python 3, then bytes (native string type) must be expected in Python 2. So we shouldn't rely on the native type, at least this is my understanding of his comment.

So I reworked the patch to follow exarkun's comment and marked each string literal with a particular type. It may look like overkill, but what is a guiding line?

> * Please create separate patch files: At least one for whitespace changes only, one for fixes independent of PY3 (like test_newclient around line 610 or changed comments further down), and one with only real PY3 changes. This one should be as small as possible.

I think changes in comments has to be part of a single patch, because things like `@type C{bytes}` directly relate to porting efforts. Other than that there are just a couple of typos that I corrected.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

> shouldn't the news file be in twisted/web/topfiles?

Correct.

> Why must HTTParser States be unicode instead of bytes? Or would it be possible to just define STATUS, HEADER, BODY, DONE = range(4)?

When trying to understand what's happening, it's much better if you see that the state is `HEADER` instead of `1`.

> I do not think getattr() should get the attribute name as unicode.

Correct.  The type of the 2nd argument to `getattr` is "native string" which is what the `""` (and `''`) literal syntaxes give you.  It is correct to leave these uses of the string literal syntax.

> Same for DEFERRED_CLOSE and friends.

Correct.

> Values for producerState: This is just an enum, no need for unicode (see HTTParser State).

These values are text so unicode is a sensible type for them.  However, mixing unicode and bytes is a bad idea.  If these tests depend on `proto_helpers.py`, possibly that module should be ported first.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

> Then exarkun commented on the ticket, see comment:5, telling that relying on the native string type gives no indication that someone considered a particular type of the string, and also it is better when API supports exactly one type of the string in both versions of Python. I.e. if native string (text) is expected in Python 3, then unicode must be expected in Python 2. If bytes are expected in Python 3, then bytes (native string type) must be expected in Python 2. So we shouldn't rely on the native type, at least this is my understanding of his comment.

Sorry if I confused you on this point.  In general what you've summarized here is correct.  However, the exception is where an API provided by someone else (either Python itself or a third-party) has chosen that the type for some API is bytes on Python 2 and unicode on Python 3.  Then we have to go along with that decision.

**Comment by twisted-trac** [comment]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|wolfgang61 commented|
|-|-|

Replying to [maximlacrima](#note_9):
> Replying to [wolfgang61](#note_8):
> > * Please create separate patch files: At least one for whitespace changes only, one for fixes independent of PY3 (like test_newclient around line 610 or changed comments further down), and one with only real PY3 changes. This one should be as small as possible.
> 
> I think changes in comments has to be part of a single patch, because things like `@type C{bytes}` directly relate to porting efforts. Other than that there are just a couple of typos that I corrected.

In general the shorter the patch the easier it is to review. That is why I would prefer to split things not absolutely needed for what a ticket is about into separate patch files. If the important changes make a line too long, I personally would even put splitting that line into such a "preparation" patch, but others might think this goes too far. Of course still leaving all patches within the same ticket.

With whitespace changes I meant things like _newclient.py around line 475 where you break a long line without any other changes.

**Comment by twisted-trac** [comment]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|wolfgang61 commented|
|-|-|

Replying to [exarkun](#note_10):
> > Why must HTTParser States be unicode instead of bytes? Or would it be possible to just define STATUS, HEADER, BODY, DONE = range(4)?
> 
> When trying to understand what's happening, it's much better if you see that the state is `HEADER` instead of `1`.

Sure, that was just trying to find out what would break if moving away from unicode. But anyway - why unicode? Those are fixed ascii-only values which never change and are never transferred to another process. I would rather use unicode only when there is a real reason to do so. One could also argue that such values which are only used internally and never cross process borders should be native strings. Which I would prefer because it makes the patch smaller.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

> I would rather use unicode only when there is a real reason to do so.

The correct types should always be used.  These values are text.  Perhaps it would be more clear that they're text if they were being displays to a user somewhere - but the values are only there for human consumption.  That's what makes them text.  The correct type for text is `unicode`.

It would be better if these were defined using `twisted.python.constants`, though.  `NamedConstant` is a better type than either bytes or text for these.  Initially I didn't make this comment because I thought `twisted.python.constants` had not yet been ported, but now I see that it has been.

It's true that this patch is much larger than the ideal patch.  Fixing these constants in a separate contribution would make sense to me.

**Comment by twisted-trac** [comment]
|<img alt="maximlacrima's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|maximlacrima commented|
|-|-|

New patch:

* Native string type is used in the following cases:
  * if string literal is passed directly or indirectly to `getattr`. This includes:
    * the first argument to `_newclient.makeStatefulDispatcher`.
    * `_newclient.Response._state` assignments
    * `_newclient.HTTP11ClientProtocol._state` assignments
  * in comparison operations, member tests etc. in which `_state` attribute is involved.
  * if string literal is passed to `t.p.c.networkString`.
* In all other cases string literals are explicitly marked with `b` or `u`.
* news file moved to twisted/web/topfiles.
* the patch doesn't contain whitespace changes on its own. There are two unrelated to Python 3 changes, these are typos corrected at _newclient.py, line 941 and test_newclient.py, line 2613.  Other changes are all related to Python3 porting.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun removed owner|
|-|-|

Submitted for review by maximlacrima so probably not being reviewed by maximlacrima.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph commented|
|-|-|

(In [43672]) Branching to newclient-py3-7485.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph set owner to @glyph|
|-|-|

The patch looks like it's in good shape to me; all previous comments appear to be addressed.  It _is_ very long, so my eyes necessarily glazed over a bit, but the test automation is pretty happy and the types seem to be right as per the discussion above.

My two points of feedback are the one new "line too long" warning that the buildbot found, and that I will be making the NEWS file into a .misc.  While this change was a substantial amount of work, it still isn't quite user-facing: `_newclient` is an internal implementation module, and not all of `Agent` is ported yet, so there's no point in putting this in the changelog.  That doesn't mean we don't value this change!  It gets us a _lot_ close to `Agent` itself being ported.

I'll address these and merge.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph commented|
|-|-|

Oh, I guess the NEWS file was a `.misc`, it just wasn't empty as is the custom :).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph set status to `closed`|
|-|-|

(In [43677]) Merge newclient-py3-7485: Port twisted.web._newclient to Python 3.

Author: maximlacrima

Reviewer: glyph

Fixes: [#7485](https://github.com/twisted/twisted/issues/7485)

Most of the internals of twisted.web.client.Agent now work on python 3,
although Agent itself still needs to be ported.
